components/core/tests/test-ir_serializer.cpp

```diff
@@ -0,0 +1,147 @@
+using clp::streaming_compression::zstd::Decompressor;
+
+namespace {
+template <typename encoded_variable_t>
+auto match_encoding_type(bool is_four_bytes_encoding) -> bool {
+    static_assert(
+            (is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>)
+            || (is_same_v<encoded_variable_t, four_byte_encoded_variable_t>)
+    );
+
+    if constexpr (is_same_v<encoded_variable_t, eight_byte_encoded_variable_t>) {
+        return false == is_four_bytes_encoding;
+    } else {
+        return is_four_bytes_encoding;
+    }
+}
+}  // namespace
+
+/*
+ * The test case only covers four byte encoding because decompressor
+ * does not support eight bytes encoding yet.
+ */
+TEMPLATE_TEST_CASE(
+        "Serialize log events",
+        "[ir][serialize-log-event]",
+        four_byte_encoded_variable_t
+) {
+    LogEventSerializer<TestType> serializer;
+
+    // Test encoding with serializer
+    vector<string> var_strs
+            = {"4938",
+               std::to_string(INT32_MAX),
+               std::to_string(INT64_MAX),
+               "0.1",
+               "-25.519686",
+               "-25.5196868642755",
+               "-00.00",
+               "bin/python2.7.3",
+               "abc123"};
+    size_t var_ix{0};
+
+    string first_log_event = "here is first string with a small int " + var_strs[var_ix++];
+    auto const first_ts
+            = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+    first_log_event += " and a medium int " + var_strs[var_ix++];
+    first_log_event += " and a very large int " + var_strs[var_ix++];
+    first_log_event += " and a small float " + var_strs[var_ix++];
+    first_log_event += "\n";
+
+    auto second_log_event = "here is second string with a medium float " + var_strs[var_ix++];
+    auto const second_ts
+            = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+    second_log_event += " and a weird float " + var_strs[var_ix++];
+    second_log_event += " and a string with numbers " + var_strs[var_ix++];
+    second_log_event += " and another string with numbers " + var_strs[var_ix++];
+    second_log_event += "\n";
+
+    string ir_test_file = "ir_serializer_test";
+    ir_test_file += cIrFileExtension;
+
+    REQUIRE(serializer.open(ir_test_file));
+    // Test serializing log events
+    REQUIRE(serializer.serialize_log_event(first_ts, first_log_event));
+    REQUIRE(serializer.serialize_log_event(second_ts, second_log_event));
+    serializer.close();
+
+    Decompressor ir_reader;
+    ir_reader.open(ir_test_file);
+
+    bool uses_four_byte_encoding{false};
+    REQUIRE(IRErrorCode_Success
+            == clp::ffi::ir_stream::get_encoding_type(ir_reader, uses_four_byte_encoding));
+    REQUIRE(true == uses_four_byte_encoding);
+
+    auto result = LogEventDeserializer<TestType>::create(ir_reader);
+    REQUIRE(false == result.has_error());
+    auto deserializer_inst = std::move(result.value());
```

Comment:
```suggestion
    auto& deserializer_inst = result.value();
```
Can we just hold a reference?

Reply:
I don't think move is technically wrong here, since we are not going to use results.value anymore. is there a reason why you prefer reference other than performance?

Reply:
It's not technically wrong. But the result and the underlying deserializer here should have the same life cycle as they r defined and used in the same scope. There's no reason to explicitly call `move` which depends on whether the underlying deserializer has move semantics supported. It's like using `std::optional`, where you just use `.value()` to access the value instead of calling `move`, right?

Reply:
Sure, the same scope part persuades me